engine/jobmaster/dm/ddl_coordinator.go

```diff
@@ -0,0 +1,812 @@
+				return true
+			}
+		}
+		if cmp, err := joined.Compare(postTable); err != nil || cmp < 0 {
+			continue
+		}
+
+		return true
+	}
+	return false
+}
+
+func (g *shardGroup) allTableSmaller(tp tableType) bool {
+	var (
+		joined schemacmp.Table
+		err    error
+	)
+	joined, err = g.joinTables(tp)
+
+	if err != nil {
+		return false
+	}
+
+	for sourceTable := range g.conflictTables {
+		t := genCmpTable(g.normalTables[sourceTable])
+		if _, err = joined.Compare(t); err == nil {
+			return false
+		}
+	}
+	return true
+}
+
+func (g *shardGroup) allTableLarger(tp tableType) bool {
+	for sourceTable, conflictTableStmt := range g.conflictTables {
+		conflictTable := genCmpTable(conflictTableStmt)
+		// for every conflict table's prev_table
+		normalTable := genCmpTable(g.normalTables[sourceTable])
+
+		for s := range g.normalTables {
+			// for every judge table
+			judgeTableStmt, ok := g.getTableBySourceTable(s, tp)
+			if !ok || judgeTableStmt == "" {
+				continue
+			}
+			judgeTable := genCmpTable(judgeTableStmt)
+
+			joined, err := normalTable.Join(judgeTable)
+			if err != nil {
+				// modify column
+				if _, err := judgeTable.Join(conflictTable); err != nil {
+					return false
+				}
+			} else if cmp, err := joined.Compare(conflictTable); err != nil || cmp < 0 {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (g *shardGroup) resolveTables() {
+	for sourceTable, conflictStmt := range g.conflictTables {
+		g.normalTables[sourceTable] = conflictStmt
+	}
+	g.conflictTables = make(map[metadata.SourceTable]string)
+	// TODO: redirect for conflict worker.
+}
+
+func (g *shardGroup) getTableForOneDDL(item *metadata.DDLItem, idx int) (string, string) {
+	return item.Tables[idx], item.Tables[idx+1]
+}
+
+func (g *shardGroup) getTableBySourceTable(st metadata.SourceTable, tp tableType) (string, bool) {
+	var (
+		stmt string
+		ok   bool
+	)
+	switch tp {
+	case normal:
+		stmt, ok = g.normalTables[st]
+	case conflict:
+		stmt, ok = g.conflictTables[st]
+	case final:
+		stmt, ok = g.conflictTables[st]
+		if !ok {
+			stmt, ok = g.normalTables[st]
+		}
+	}
+	return stmt, ok
+}
+
+func (g *shardGroup) checkAddDroppedColumn(ctx context.Context, sourceTable metadata.SourceTable, ddl string, prevTableStmt, postTableStmt string, newDroppedColumns []string) (string, error) {
+	currTable := g.normalTables[sourceTable]
+	defer func() {
+		g.normalTables[sourceTable] = currTable
+	}()
+
+	g.normalTables[sourceTable] = prevTableStmt
+	oldJoined, err := g.joinTables(normal)
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	postTable := genCmpTable(postTableStmt)
+	g.normalTables[sourceTable] = postTableStmt
+	newJoined, err := g.joinTables(normal)
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	cmp, err := oldJoined.Compare(newJoined)
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	if cmp <= 0 {
+		if col, err2 := optimism.AddDifferentFieldLenColumns("", ddl, oldJoined, newJoined); err2 != nil {
```

Comment:
should we mention that checkAddDroppedColumn will also check field length?

And why this case is "add column with a larger field len"? it can be anything that modifying the table structure with the only longest field length column

Reply:
I don't want to change the original name😂